### DISCUSSION TITLE
modify raw body in api route

### DISCUSSION THREAD

**Discussion by macrozone** [vercel/next.js#27702] [Help] [answered]
I need to patch incoming graphql requests (due to a missdeployement). 

I tried  to write a middleware to change the body of the incoming request.

Problem is though, that body parser needs to be disabled, so i have to work with the raw stream.

I can read this stream and convert it to a string and then modify the string, but i don't know how i can modify the request so that my altered string is read.

I tried to override the `req.read` function, but had no success with that. Any other idea?

**Comment by jesper-bylund** [comment]
There is a hidden config for this, just export a config object like this:
```
export const config = {
  api: {
    bodyParser: false,
  },
};
```

I found this on:
https://vercel.com/guides/getting-started-with-nextjs-typescript-stripe#step-3:-handling-webhooks-&-checking-their-signatures

**Comment by leerob** [comment, marked as answer]
```js
import type { NextApiRequest, NextApiResponse } from 'next';
import type { Readable } from 'node:stream';

export const config = {
  api: {
    bodyParser: false,
  },
};

async function buffer(readable: Readable) {
  const chunks = [];
  for await (const chunk of readable) {
    chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : chunk);
  }
  return Buffer.concat(chunks);
}

export default async function (req: NextApiRequest, res: NextApiResponse) {
  if (req.method === 'POST') {
    const buf = await buffer(req);
    const rawBody = buf.toString('utf8');

    res.json({ rawBody });
  } else {
    res.setHeader('Allow', 'POST');
    res.status(405).end('Method Not Allowed');
  }
}
```

[Source](https://vercel.com/support/articles/how-do-i-get-the-raw-body-of-a-serverless-function)

**Reply by mglind** [reply]
What if you just want to intercept the raw body so you can do something with it and then pass the request along with the buffer in tact? This example seems to get the body fine but it empties the buffer so passing the request along to something else no longer has this readable buffer that it was expecting.

**Reply by activenode** [reply]
This indeed works

**Reply by darklight9811** [reply]
Is this going to be supported on route handlers? (app dir)

**Reply by leerob** [reply]
```
await request.text()
```

**Reply by asongate** [reply]
Thank you @leerob!
I was finally able to get the raw body in the route handler ( app dir ) using `await request.text() `

**Reply by marcelo-albuquerque** [reply]
> ```
> await request.text()
> ```

Thank you @leerob! This is the perfect solution, simple and functional.

To convert the Raw Body to JSON without errors:

```ts
export async function POST(request: Request) {
  const rawBody = await request.text();
  const jsonBody = JSON.parse(rawBody);

  return Response.json(jsonBody);
}
```
